ndk/src/thermal.rs

```diff
@@ -0,0 +1,331 @@
+#![cfg(feature = "api-level-30")]
+
+#[cfg(doc)]
+use std::io::ErrorKind;
+use std::{io::Result, os::raw::c_void, ptr::NonNull};
+
+use num_enum::{FromPrimitive, IntoPrimitive};
+
+use crate::utils::abort_on_panic;
+
+/// Workaround for <https://issuetracker.google.com/issues/358664965>.  `status_t` should only
+/// contain negative error codes, but the underlying `AThermal` implementation freely passes
+/// positive error codes around. At least the expected return codes are "implicitly" documented to
+/// be positive.
+fn status_to_io_result(status: i32) -> Result<()> {
+    // Intentionally not imported in scope (and an identically-named function) to prevent
+    // accidentally calling this function without negation.
+    crate::utils::status_to_io_result(-status)
+}
+
+/// Thermal status used in function [`ThermalManager::current_thermal_status()`] and
+/// [`ThermalStatusCallback`].
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, FromPrimitive, IntoPrimitive)]
+#[repr(i32)]
+#[doc(alias = "AThermalStatus")]
+#[non_exhaustive]
+pub enum ThermalStatus {
+    /// Error in thermal status.
+    // TODO: Move to a Result?
+    #[doc(alias = "ATHERMAL_STATUS_ERROR")]
+    Error = ffi::AThermalStatus::ATHERMAL_STATUS_ERROR.0,
+    /// Not under throttling.
+    #[doc(alias = "ATHERMAL_STATUS_NONE")]
+    None = ffi::AThermalStatus::ATHERMAL_STATUS_NONE.0,
+    /// Light throttling where UX is not impacted.
+    #[doc(alias = "ATHERMAL_STATUS_LIGHT")]
+    Light = ffi::AThermalStatus::ATHERMAL_STATUS_LIGHT.0,
+    /// Moderate throttling where UX is not largely impacted.
+    #[doc(alias = "ATHERMAL_STATUS_MODERATE")]
+    Moderate = ffi::AThermalStatus::ATHERMAL_STATUS_MODERATE.0,
+    /// Severe throttling where UX is largely impacted.
+    #[doc(alias = "ATHERMAL_STATUS_SEVERE")]
+    Severe = ffi::AThermalStatus::ATHERMAL_STATUS_SEVERE.0,
+    /// Platform has done everything to reduce power.
+    #[doc(alias = "ATHERMAL_STATUS_CRITICAL")]
+    Critical = ffi::AThermalStatus::ATHERMAL_STATUS_CRITICAL.0,
+    /// Key components in platform are shutting down due to thermal condition. Device
+    /// functionalities will be limited.
+    #[doc(alias = "ATHERMAL_STATUS_EMERGENCY")]
+    Emergency = ffi::AThermalStatus::ATHERMAL_STATUS_EMERGENCY.0,
+    /// Need shutdown immediately.
+    #[doc(alias = "ATHERMAL_STATUS_SHUTDOWN")]
+    Shutdown = ffi::AThermalStatus::ATHERMAL_STATUS_SHUTDOWN.0,
+
+    #[doc(hidden)]
+    #[num_enum(catch_all)]
+    __Unknown(i32),
+}
+
+impl From<ffi::AThermalStatus> for ThermalStatus {
+    fn from(value: ffi::AThermalStatus) -> Self {
+        value.0.into()
+    }
+}
+
+/// Prototype of the function that is called when thermal status changes. It's passed the updated
+/// thermal status as parameter.
+///
+/// # Warning
+/// [`ThermalManager`] is synchronized internally, and its lock is held while this callback is
+/// called.  Interacting with [`ThermalManager`] inside this closure *will* result in a deadlock.
```

Comment:
While all the extra locking makes it possible to add `Send`+`Sync` on `ThermalManager`, it means the user can now trivially pass this type into the callback and (if it's the same instance!) get a deadlock when interacting with it.